### ISSUE TITLE
ugly error message for matrix*JuMPDict

### ISSUE THREAD

**Issue by mlubin** [JuliaOpt/JuMP.jl#568]
On 0.4:
```
julia> m = Model();

julia> B = rand(2,2);

julia> @defVar(m, x[[1,2]]);

julia> @addConstraint(m, B*x .== [3,4])
ERROR: MethodError: `_sz` has no method matching _sz(::Array{Float64,2}, ::JuMP.JuMPDict{JuMP.Variable,1})
Closest candidates are:
  _sz(::AbstractArray{T,N}, ::Union{JuMP.AbstractJuMPScalar,JuMP.GenericAffExpr{CoefType,VarType},JuMP.GenericNormExpr{2,Float64,JuMP.Variable},JuMP.GenericNorm{P,Float64,JuMP.Variable},JuMP.GenericQuadExpr{Float64,JuMP.Variable},Number})
  _sz(::AbstractArray{T,N}, ::AbstractArray{T,2})
  _sz(::AbstractArray{T,2}, ::AbstractArray{T,1})
 in addToExpression at /home/mlubin/.julia/v0.4/JuMP/src/v0.4/parseExpr_staged.jl:303
 in addToExpression_reorder at /home/mlubin/.julia/v0.4/JuMP/src/v0.4/parseExpr_staged.jl:316
```

**Comment by joehuchette** [comment]
This only occurs inside ``@addConstraint``, and also produces the same error for JuMPArrays.

**Comment by joehuchette** [comment]
Only tangentially related, but I think there should be a way to remove most of the vectorized operator code (except matmul) by using ``promote_type``-like stuff, but I'm pretty sure it'd only work on v0.4. In the meantime, we can just have a better error message here.

**Comment by joehuchette** [comment]
Closed with #576

**Comment by mlubin** [comment]
Now the error message is:
```
ERROR: MethodError: `*` has no method matching *(::Array{Float64,2}, ::JuMP.JuMPDict{JuMP.Variable,1})
Closest candidates are:
  *(::Any, ::Any, ::Any, ::Any...)
  *{T<:Union{JuMP.AbstractJuMPScalar,JuMP.GenericAffExpr{CoefType,VarType},JuMP.GenericNormExpr{2,Float64,JuMP.Variable},JuMP.GenericNorm{P,Float64,JuMP.Variable},JuMP.GenericQuadExpr{Float64,JuMP.Variable}}}(::Union{Array{T,2},Base.SparseMatrix.SparseMatrixCSC{Tv,Ti<:Integer}}, ::Union{Array{T<:Union{JuMP.AbstractJuMPScalar,JuMP.GenericAffExpr{CoefType,VarType},JuMP.GenericNormExpr{2,Float64,JuMP.Variable},JuMP.GenericNorm{P,Float64,JuMP.Variable},JuMP.GenericQuadExpr{Float64,JuMP.Variable}},1},Array{T<:Union{JuMP.AbstractJuMPScalar,JuMP.GenericAffExpr{CoefType,VarType},JuMP.GenericNormExpr{2,Float64,JuMP.Variable},JuMP.GenericNorm{P,Float64,JuMP.Variable},JuMP.GenericQuadExpr{Float64,JuMP.Variable}},2},Base.SparseMatrix.SparseMatrixCSC{T<:Union{JuMP.AbstractJuMPScalar,JuMP.GenericAffExpr{CoefType,VarType},JuMP.GenericNormExpr{2,Float64,JuMP.Variable},JuMP.GenericNorm{P,Float64,JuMP.Variable},JuMP.GenericQuadExpr{Float64,JuMP.Variable}},Ti<:Integer}})
  *{T<:Union{Complex{Float32},Complex{Float64},Float32,Float64},S}(::Union{DenseArray{T<:Union{Complex{Float32},Complex{Float64},Float32,Float64},2},SubArray{T<:Union{Complex{Float32},Complex{Float64},Float32,Float64},2,A<:DenseArray{T,N},I<:Tuple{Vararg{Union{Colon,Int64,Range{Int64}}}},LD}}, ::Union{DenseArray{S,1},SubArray{S,1,A<:DenseArray{T,N},I<:Tuple{Vararg{Union{Colon,Int64,Range{Int64}}}},LD}})
  ...
 in addToExpression_reorder at /home/mlubin/.julia/v0.4/JuMP/src/v0.4/parseExpr_staged.jl:257
```
Not super friendly either
